Update an image and replace screenshots (#148)
neo4j/docs-getting-started: modules/ROOT/pages/cypher-intro/large-statements.adoc

[[cypher-intro-large-statements-with]]
== WITH

> In Cypher, you are able to chain fragments of statements together, similar to how it is done within a data-flow pipeline.

Reviewer: It's better to not use personal pronouns wherever it is possible. But if you have to use them, I'd say something like:
_In Cypher, you can chain fragments of statements together...."_

> In Cypher, you are able to chain fragments of statements together, similar to how it is done within a data-flow pipeline.

Reviewer: ```suggestion
In Cypher, it is possible to chain fragments of statements together, similar to how it is done within a data-flow pipeline.
```

> Each fragment works on the output from the previous one, and its results can feed the next one.

Reviewer: I don't think you should leave out _into_ in this case.

> Each fragment works on the output from the previous one, and its results can feed the next one.

Reviewer: ```suggestion
Each fragment works on the output from the previous one, and its results can feed into the next one.
```

Expressions, aggregations, ordering and pagination can be used in the same way as in the `RETURN` clause.
The only difference is all columns must be aliased.

> In the following example, you first find the collection of movies someone appeared in and then filter out those which appear in only one movie.

Reviewer: ```suggestion
In the following example, collect the movies someone appeared in and then filter out those which appear in only one movie.
```

+-------------------------------------------------------------+
----

Using the `WITH` clause, you can pass values from one section of a query to another.
This allows you to perform some intermediate calculations or operations within your query to use later.

> Take a look at the following examples based on this dataset:

Reviewer: ```suggestion
The following examples are based on this dataset:
```

image:people-technologies-graph-arr.svg[role="popup-link"]

You must specify the variables in the `WITH` clause that you want to use later.

> Only those variables are to be passed to the next part of the query.

Reviewer: ```suggestion
Only those variables are passed on to the next part of the query.
```

| 'Ann'       | ['Application Development']                |
| 'Mark'      | ['Graphs']                                 |
| 'Joe'       | ['Query Languages']                        | 
| 'Melissa'   | ['Query Languages', 'Data ETL', 'Graphs']  |
+----------------------------------------------------------+
----

[source, cypher]
----
//Query2: find number of friends who have other friends
MATCH (p:Person)-[:IS_FRIENDS_WITH]->(friend:Person)
WITH p, collect(friend.name) AS friendsList, count{(friend)-[:IS_FRIENDS_WITH]-(:Person)} AS numberOfFoFs
WHERE numberOfFoFs > 1
RETURN p.name, friendsList, numberOfFoFs;
----

*Query2 results:*
[queryresult]
----
Rows: 3

+---------------------------------------------------+-----------------+
| p.name        | friendList                        | numberOfFoFs    |     
+---------------------------------------------------+-----------------+
| 'Joe'         | ['Mark']                          | 2               |       
| 'Jennifer'    | ['Sally', 'John', 'Ann', 'Mark']  | 2               |
| 'John'        | ['Sally']                         | 2               |
+---------------------------------------------------+-----------------+
----

> In the first query, the `Person` name and a collected list of the `Technology` types are being passed.

Reviewer: ```suggestion
In the first query, the `Person` name and a collected list of the `Technology` types are passed.
```

> Therefore only these items can be referenced in the `RETURN` clause.

Reviewer: ```suggestion
Therefore, only these items can be referenced in the `RETURN` clause.
```

> You cannot use the relationship (`r`) or even the `Person` birthdate because you did not pass those values along.

Reviewer: ```suggestion
Neither the relationship (`r`) nor the `Person` birthdate can be used because those values were not passed along.
```

> In the second query, you can reference `p` and any of its properties (`name`, `birthdate`, `yearsExperience`, `twitter`), the collection of friends (as a whole, not each value), and the number of friend-of-friends.

Reviewer: ```suggestion
In the second query, only `p` and any of its properties (`name`, `birthdate`, `yearsExperience`, `twitter`), the collection of friends (as a whole, not each value), and the number of friend-of-friends can be referenced.
```

> Since you passed those values in the `WITH` clause, you can use those in `WHERE` or `RETURN` clauses.

Reviewer: ```suggestion
Since those values were passed along in the `WITH` clause, those can be used in `WHERE` or `RETURN` clauses.
```

`WITH` requires all values passed to have a variable (if they do not already have one).

> The `Person` nodes are given a variable (`p`) in the `MATCH` clause, so you do not need to assign a variable there.

Reviewer: ```suggestion
The `Person` nodes are given a variable (`p`) in the `MATCH` clause, so no variable needs to be assigned there.
```
